### PR TITLE
Padding free squashing

### PR DESCRIPTION
**Pull request by garrett361**
Squashes various commits and rebases on top of latest upstream main.

### CODE REVIEW

**Comment by fabianlim** [comment]
@garrett361 i didnt realize there is a quality check

File: open_instruct/finetune.py [hunk 1]
```diff
@@ -630,10 +682,19 @@ def main(args: FlatArguments, tc: TokenizerConfig):
         model.gradient_checkpointing_enable()
 
     # DataLoaders creation:
+    if args.padding_free:
+        accelerator.print("Using padding-free collation")
```
Comment (fabianlim) [inline open_instruct/finetune.py:686]: if print statements are not needed maybe we can remove

File: open_instruct/finetune.py [hunk 2]
```diff
@@ -848,6 +1011,7 @@ def main(args: FlatArguments, tc: TokenizerConfig):
                         logger.info(
                             f"  Step: {completed_steps}, LR: {lr_scheduler.get_last_lr()[0]}, Loss: {avg_loss}, TPS: {total_tokens / (time.time() - start_time)}"
                         )
+                    accelerator.print(f"{metrics_to_log=}")
```
Comment (fabianlim) [inline open_instruct/finetune.py:1014]: same as above

File: tests/test_padding_free.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            **{
+                **model_kwargs,
+                "torch_dtype": self.dtype,
+                "attn_implementation": "flash_attention_2",
+            }
+        )
+        model = model_cls(cfg).to("cuda", dtype=self.dtype)
+        return model, cfg
+
+    @pytest.mark.parametrize("model_name", ["bamba", "llama"])
```
Comment (fabianlim) [inline tests/test_padding_free.py:81]: do your tests pass?

**Comment by garrett361** [comment]
@fabianlim also didn't realize about the check; guess it's only activated when PR-ing against main.

File: open_instruct/dataset_processor.py [hunk 4]
```diff
@@ -166,6 +166,60 @@
         "{% endif %}"
         "{% endfor %}"
     ),
+    "granite": (
```
[the remark this answers is not in the document]
Comment (garrett361) [reply open_instruct/dataset_processor.py:169]: Divya required this for her work. Easy to drop this later since it's a standalone PR.

File: open_instruct/finetune.py (hunk 1 above)
Comment (garrett361) [reply open_instruct/finetune.py:686]: oh yeah, thought I had done so, will do

File: open_instruct/finetune.py (hunk 2 above)
Comment (garrett361) [reply open_instruct/finetune.py:1014]: This one is super useful, though, so I'm going to keep it. It's part of the `extra reporting`  commit which is really just for dev testing and not intended to ever be upstreamed.

File: tests/test_padding_free.py (hunk 3 above)
Comment (garrett361) [reply tests/test_padding_free.py:81]: Testing in `us.icr.io/cil15-shared-registry/platform/goon-open-instruct-flash-mamba:latest`, I'm finding `pytest test_padding_free.py` passes.

**Comment by fabianlim** [comment]
I think this will be a good to merge soon, although we need to document down what is missing in comparison to[ padding-free branch](https://github.com/garrett361/open-instruct/tree/padding-free)
1. accept a `train_file` input
2. parquet file support and additional model args https://github.com/garrett361/open-instruct/pull/2
3. tools and documents (is this still needed)? https://github.com/garrett361/open-instruct/pull/3
4. dataloader tester https://github.com/garrett361/open-instruct/pull/7
5. `add_column` is slow so use pandas to load dataset https://github.com/garrett361/open-instruct/pull/6/files

File: open_instruct/dataset_processor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (garrett361) [reply open_instruct/dataset_processor.py:169]: Yeah, I think I'll do 2: if `chat_template_name` isn't in `CHAT_TEMPLATES`, then try to read in the text from a file at path `chat_template_name`. Kinda hacky, but then we don't have to add more flags.

[the remark this answers is not in the document]
Comment (fabianlim) [reply open_instruct/dataset_processor.py:169]: hmm.. but maybe 1. is better because its more consistent..

**Comment by fabianlim** [comment]
this seems to be superceeded by the series of PRs #17 , etc.. so maybe we should close it?
